### PR TITLE
Fix state_quant to preserve the input tensor's dtype

### LINKED ISSUE #439: state_quant returns float32 for float64 inputs
**Issue type**

- [x] Bug Report
- [ ] Feature Request
- [ ] Help wanted
- [ ] Other

**snnTorch version**

`1.0.0`

**Description**

I found a reproducible dtype bug in the public helper `snntorch.functional.quant.state_quant`.

When the input tensor uses `torch.float64`, the quantized output is returned as `torch.float32`. This occurs with both uniform and non-uniform quantization.

As a result, applying `state_quant` to a double-precision hidden state silently changes its dtype. This can cause precision loss or dtype mismatches in a model converted with `.double()`.

From source inspection, the quantization levels are created using `torch.linspace(...)` or `torch.tensor(...)` without specifying a dtype. In `StateQuant.forward`, the levels are moved to the input device with `levels.to(device)`, but they are not converted to the input dtype. The returned tensor is then selected directly from those levels, so it inherits their default `float32` dtype.

I reproduced this with both snnTorch `1.0.0` and the current master commit in fresh CPU-only Docker containers using Debian GNU/Linux 12, Python 3.11.16, and PyTorch 2.8.0+cpu.

**Minimal code to reproduce the error/bug**

```python
import torch
from snntorch.functional import quant

torch.set_default_dtype(torch.float32)

x = torch.tensor([0.125, 0.875], dtype=torch.float64)

uniform_quant = quant.state_quant(
    num_bits=4,
    uniform=True,
    threshold=1.0,
)
nonuniform_quant = quant.state_quant(
    num_bits=4,
    uniform=False,
    thr_centered=False,
    threshold=1.0,
)

uniform_output = uniform_quant(x)
nonuniform_output = nonuniform_quant(x)

print("input_dtype =", x.dtype)
print("uniform_output_dtype =", uniform_output.dtype)
print("nonuniform_output_dtype =", nonuniform_output.dtype)
```

**What is expected**

`state_quant` should preserve the floating-point dtype of the input tensor.

For a `torch.float64` input, both quantized outputs should also use `torch.float64`.

**What actually happens**

Both uniform and non-uniform quantization return `torch.float32` tensors for a `torch.float64` input.

Example output from the Docker reproduction:

```text
input_dtype = torch.float64
uniform_output_dtype = torch.float32
nonuniform_output_dtype = torch.float32
```

### PR DESCRIPTION
**Pull request by li-lizhe**
### Problem

`snntorch.functional.quant.state_quant` silently returns `float32` tensors for `float64` inputs, with both uniform and non-uniform quantization (#439). This causes precision loss or dtype mismatches when quantizing the double-precision hidden states of a model converted with `.double()`.

### Root cause

The quantization levels are created with `torch.linspace(...)` / `torch.tensor(...)` without a dtype, so they default to `float32`. In `StateQuant.forward`, the levels are only moved to the input device:

```python
levels = levels.to(device)
```

They are never converted to the input dtype, and the returned tensor is selected directly from those levels (`levels[idx_match]`), so it inherits the hard-coded `float32`.

### Fix

Convert the levels to the input dtype while moving them to the input device:

```python
levels = levels.to(device=device, dtype=input_.dtype)
```

Dtype-agnostic and device-agnostic: no-op for `float32` inputs (today's behavior), preserves `float64` (and any other floating dtype) everywhere.

### Verification

* Issue reproduction: a `float64` input now returns `float64` for uniform, non-uniform `thr_centered=True`, and non-uniform `thr_centered=False` quantization.
* Quantized values are unchanged vs `float32` inputs (same nearest levels selected); straight-through-estimator gradients still flow with the input dtype (`grad.dtype == float64`).
* New regression tests in `tests/test_snntorch/functional/test_quant_dtype.py` (3 quantization modes x {float32, float64}).
* Cross-device check on Ascend NPU (torch 2.14.0a0 + torch_npu, 910B): dtype preserved on `npu:0` for `float32`. (`float64` on Ascend is limited by torch_npu's `aclnnMinDim` dtype support — identical behavior before and after this change, i.e. not introduced here.)

Fixes #439

### CODE REVIEW

**Comment by ixfd64** [comment]
Your changes are causing the automated tests to fail. Please take a look at them.

**Comment by li-lizhe** [comment]
Thanks for the heads-up! I found the root cause: the test file `test_quant_dtype.py` was shared between two PRs — #444 (state_quant fix) and #445 (latency fix). #444 had the `test_latency_preserves_dtype` test but without the latency fix ([included in #445](https://github.com/jeshraghian/snntorch/pull/445)), causing CI to fail on the latency test. I've removed the latency test from #444 and pushed the fix. #444 now only tests `test_state_quant_preserves_dtype`. The latency dtype preservation is correctly tested in #445. CI should pass now.
